README.md
# RPi-Band

RPi-Band fuses the Pimoroni Piano HAT and Drum HAT code to a single software,
which gives the possibility for playing two intruments on a single Raspberry Pi.
The code in this repository is Python3, but rewriting for Python2 should be no issue.

# Tutorial
Find more info in my tutorial  
http://www.knight-of-pi.org/rpi-band-drumhat-and-pianohat-simultaneously-on-a-single-raspberry-pi/

# Hardware setup
![Alt text](rpi_band_setup.jpg?raw=true "RPi-Band")

# Installing RPi-Band

    cd ~
    sudo apt-get install python3-pianohat
    sudo apt-get install python3-drumhat
    git clone https://github.com/JoBergs/RPi-Band


# Usage

    cd ~/RPi-band
    python3 rpi-band.py

# Autostart
Opening .bashrc with

    sudo nano ~/.bashrc
and add the line

    python3 /home/pi/RPi-Band/rpi-band.py

# Parent GitHub repositories and original READMEs 
Piano HAT: https://github.com/pimoroni/Piano-HAT  
Drum HAT: https://github.com/pimoroni/Drum-HAT

# Have fun!

CHANGES.txt
v0.10 Plain merge of original Pimoroni drum and piano HAT code
v0.11 Piano and Drum HAT code now in (buggy) OOP
v0.12 Added function for setting pygame.mixer to NORMAL or 8BIT
v0.13 Recreate Piano instance everytime the instrument button is pressed
v0.14 integrated basic 8bit synthi
v0.15 switching wave types for synthi on octave up or down works

8bit-synth.py
#!/usr/bin/env python3

import math
import signal
import time
from sys import exit

try:
    import numpy
except ImportError:
    exit("This script requires the numpy module\nInstall with: sudo pip install numpy")

try:
    import pygame
except ImportError:
    exit("This script requires the pygame module\nInstall with: sudo pip install pygame")

import pianohat


print("""
8-bit Piano HAT

This advanced example demonstrates software synthesis.

It uses pygame's sndarray.make_sound method to create
8bit tones with specific frequency, bitrate and samplerates.

Instrument = Toggle Sine Wave
Octave ^ = Toggle Saw Wave
Octave V = Toggle Square Wave

Please wait while Pygame is set up and samples are generated...

Press CTRL+C to exit!
""")

BITRATE = 8
SAMPLERATE = 44100

ATTACK_MS=25
RELEASE_MS=500

# Fee; free to change the volume!
volume = {'sine':0.8, 'saw':0.4, 'square':0.4}

wavetypes = ['sine','saw','square']
enabled = {'sine':True, 'saw':False, 'square':False}
notes = {'sine':[],'saw':[],'square':[]}


# The samples are 8bit signed, from -127 to +127
# so the max amplitude of a sample is 127
max_sample = 2**(BITRATE - 1) - 1

def handle_instrument(channel, pressed):
    """Handles the Instrument, Octave Up/Down keys
    These toggle Sine, Saw and Square waves on and off so you can combine them
    """
    if not pressed:
        return

    if channel == 15: # Sine
        enabled['sine'] = not enabled['sine']
        print("Sine: {}".format("ON" if enabled['sine'] else "OFF"))
    if channel == 14: # Saw
        enabled['saw'] = not enabled['saw']
        print("Saw: {}".format("ON" if enabled['saw'] else "OFF"))
    if channel == 13: # Square
        enabled['square'] = not enabled['square']
        print("Square: {}".format("ON" if enabled['square'] else "OFF"))

    update_leds()


def play_sample(channel, pressed):
    """Handles the piano keys
    Any enabled samples are played, and *all* samples are turned off is a key is released
    """
    pianohat.set_led(channel, pressed)

    print(enabled)

    if pressed:
        for t in wavetypes:
            if enabled[t]:
                notes[t][channel].play(loops=-1, fade_ms=ATTACK_MS)
    else:
        for t in wavetypes:
            notes[t][channel].fadeout(RELEASE_MS)


def wave_sine(freq, time):
    """Generates a single sine wave sample"""
    s = math.sin(2*math.pi*freq*time)
    return int(round(max_sample * s ))


def wave_square(freq, time):
    """Generates a single square wave sample"""
    return -max_sample if freq*time < 0.5 else max_sample


def wave_saw(freq, time):
    """Generates a single sav wave sample"""
    s = ((freq*time)*2) - 1
    return int(round(max_sample * s)) 


def update_leds():
    """Updates the Instrument and Octave LEDs to show enabled samples"""
    pianohat.set_led(15, enabled['sine'])
    pianohat.set_led(14, enabled['saw'])
    pianohat.set_led(13, enabled['square'])


def generate_sample(frequency, volume=1.0, wavetype=None):
    """Generates a sample of a specific frequency and wavetype"""
    if wavetype is None:
        wavetype = wave_square

    sample_count = int(round(SAMPLERATE/frequency))

    buf = numpy.zeros((sample_count, 2), dtype = numpy.int8)

    for s in range(sample_count):
        t = float(s)/SAMPLERATE # Time index
        buf[s][0] = wavetype(frequency, t)
        buf[s][1] = buf[s][0] # Copy to stero channel

    sound = pygame.sndarray.make_sound(buf)
    sound.set_volume(volume) # Set the volume to balance sounds

    return sound


print("Initializing subsystem...")
pygame.mixer.pre_init(SAMPLERATE, -BITRATE, 4, 256)
pygame.mixer.init()
pygame.mixer.set_num_channels(32)


print("Generating samples...")
for f in [
        261.626,
        277.183,
        293.665,
        311.127,
        329.628,
        349.228,
        369.994,
        391.995,
        415.305,
        440.000,
        466.164,
        493.883,
        523.251
    ]:
    notes['sine'] += [generate_sample(f, volume=volume['sine'], wavetype=wave_sine)]
    notes['saw'] += [generate_sample(f, volume=volume['saw'], wavetype=wave_saw)]
    notes['square'] += [generate_sample(f, volume=volume['square'], wavetype=wave_square)]


pianohat.auto_leds(False)
pianohat.on_note(play_sample)
pianohat.on_instrument(handle_instrument)
pianohat.on_octave_up(handle_instrument)
pianohat.on_octave_down(handle_instrument)
update_leds()


print("Now, make beautiful music...")


signal.pause()


midi-piano.py
#!/usr/bin/env python3

import glob
import os
import re
import signal
import time
from sys import exit

try:
    import midi
    import midi.sequencer
except ImportError:
    exit("This script requires the midi module\nInstall with: sudo pip install midi")

try:
    import pygame
except ImportError:
    exit("This script requires the pygame module\nInstall with: sudo pip install pygame")

import pianohat


hw = midi.sequencer.SequencerHardware()

octave = 5
MIDI_CLIENT = -1
MIDI_PORT   = 0
START_PATCH = 1
BANK_SIZE   = 16

supported = ['yoshimi','SunVox']

for client in supported:
    if client in hw._clients:
        MIDI_CLIENT = hw._clients[client].client
        print('Connecting to {}'.format(client))
        break

if MIDI_CLIENT == -1:
    exit('Please install and run a supported client! :)')

class Piano():
    def __init__(self):
        self.current_patch = START_PATCH
        self.seq = midi.sequencer.SequencerWrite()
        self.seq.subscribe_port(MIDI_CLIENT, MIDI_PORT)
        self.seq.start_sequencer()
        self.select_patch(self.current_patch)
    
    def note_on(self, note, velocity=100):
        self.seq.event_write(midi.NoteOnEvent(velocity=velocity, pitch=note, tick=0), False, False, True)
        print("Note_on:{}".format(note))
    def note_off(self, note):
        self.seq.event_write(midi.NoteOffEvent(velocity=100, pitch=note, tick=0), False, False, True)
        print("Note_off:{}".format(note))

    def select_patch(self, patch):
        if patch < 0 or patch >= BANK_SIZE:
            raise ArgumentError("Invalid Patch")
        self.seq.event_write(midi.ProgramChangeEvent(tick=0, channel=0, data=[patch]), False, False, True)

    def next_patch(self):
        self.current_patch += 1
        self.current_patch %= BANK_SIZE
        self.select_patch(self.current_patch)

pianohat.auto_leds(True)
piano = Piano()

def handle_note(channel, pressed):
    velocity = 100
    if pressed:
        piano.note_on((octave * 12)  + channel, velocity)
    else:
        piano.note_off((octave * 12) + channel)

def handle_instrument(channel, pressed):
    piano.next_patch()

def handle_octave_up(channel, pressed):
    global octave
    if pressed:
        octave += 1
        print('Selected Octave: {}'.format(octave))

def handle_octave_down(channel, pressed):
    global octave
    if pressed and octave > 0:
        octave -= 1
        print('Selected Octave: {}'.format(octave))

pianohat.on_note(handle_note)
pianohat.on_octave_up(handle_octave_up)
pianohat.on_octave_down(handle_octave_down)
pianohat.on_instrument(handle_instrument)

signal.pause()


learn-to-play.py
#!/usr/bin/env python3

import os
import signal
import time
from sys import exit

try:
    import pygame
except ImportError:
    exit("This script requires the pygame module\nInstall with: sudo pip install pygame")

import pianohat


print("""
This example will teach you to play a simple melody
by lighting up the keys you must press.

Press CTRL+C to exit.
""")

SOUNDS = os.path.join(os.path.dirname(__file__), "sounds/piano")

pygame.mixer.pre_init(44100, -16, 1, 512)
pygame.mixer.init()
pygame.mixer.set_num_channels(16)

melody = [
    0, 0, 7, 7, 9, 9, 7,
    5, 5, 4, 4, 2, 2, 0
]

note = 0
starting_note = 45


def next():
    global note
    pianohat.set_led(current_note(), False)
    time.sleep(0.1)
    note += 1
    note %= len(melody)
    pianohat.set_led(current_note(), True)


def current_note():
    return melody[note]


files = [
    '39172__jobro__piano-ff-025.wav',
    '39173__jobro__piano-ff-026.wav',
    '39174__jobro__piano-ff-027.wav',
    '39175__jobro__piano-ff-028.wav',
    '39176__jobro__piano-ff-029.wav',
    '39177__jobro__piano-ff-030.wav',
    '39178__jobro__piano-ff-031.wav',
    '39179__jobro__piano-ff-032.wav',
    '39180__jobro__piano-ff-033.wav',
    '39181__jobro__piano-ff-034.wav',
    '39182__jobro__piano-ff-035.wav',
    '39183__jobro__piano-ff-036.wav',
    '39184__jobro__piano-ff-037.wav'
]

samples = [pygame.mixer.Sound(os.path.join(SOUNDS, sample)) for sample in files]

pianohat.auto_leds(False)


def handle_note(channel, pressed):
    if not channel == current_note():
        return

    if channel < len(samples) and pressed:
        print('Playing Sound: {}'.format(files[channel]))
        samples[channel].play(loops=0)
        next()


def handle_instrument(channel, pressed):
    pass


def handle_octave_up(channel, pressed):
    pass


def handle_octave_down(channel, pressed):
    pass


for x in range(16):
    pianohat.set_led(x, False)

pianohat.on_note(handle_note)
pianohat.on_octave_up(handle_octave_up)
pianohat.on_octave_down(handle_octave_down)
pianohat.on_instrument(handle_instrument)

pianohat.set_led(current_note(), True)

signal.pause()


rpi-band.py
#!/usr/bin/env python3

import argparse
import glob
import math
import numpy
import os
import pygame
import re
import signal
import subprocess
import sys
import time

import drumhat
import pianohat
import RPi.GPIO as GPIO

DESCRIPTION = '''This script integrates Pimoronis Piano HAT and Drum HAT software and gives you simple, ready-to-play instruments which use .wav files located in sounds.
The parameter -p expects the name of the directory containing the sounds that should be loaded onto the piano HAT first
and -d expects the name of the directory containing the sounds that should be loaded onto the drum HAT.
For the 8-Bit-synthesizer, pass "8bit" for -p.


Press CTRL+C to exit.'''

# accept 8bit for the synthi, extend sound_sets by it but handle specially 
SOUND_BASEDIR = os.path.join(os.path.dirname(__file__), "sounds/")

# list of all available soundsets
sound_sets = [os.path.basename(tmp) for tmp in glob.glob(os.path.join(SOUND_BASEDIR, "*"))]

sound_sets.append("8bit")

GPIO.setmode(GPIO.BCM)

# safe shutdown button is pin 14 (GND) and pin 18(IO: 24 in BCM) in BOARD numbering
GPIO.setup(24, GPIO.IN, pull_up_down=GPIO.PUD_UP)

MIXER_NORMAL = (44100, -16, 1, 512)
MIXER_8BIT = (44100, -8, 4, 256)

# needs to happen before generate_samples
def set_mixer(mixer_values):
    pygame.mixer.quit()
    pygame.mixer.pre_init(*mixer_values)
    pygame.mixer.init()
    pygame.mixer.set_num_channels(32)

# needs to be 8bit first for generating samples; is changed according to
# the chosen instrument later on
set_mixer(MIXER_8BIT)

############## synthi constants
SAMPLERATE = 44100
BITRATE = 8

ATTACK_MS = 25
RELEASE_MS = 500

# Feel free to change the volume!
volume = {'sine':0.8, 'saw':0.4, 'square':0.4}

wavetypes = ['sine','saw','square']
enabled = {'sine':True, 'saw':False, 'square':False}

# build a list of legal wave type combinations
LEGAL_WAVES = [[x, y, z] for x in [True, False] for y in [True, False] for z in [True, False]]
LEGAL_WAVES.remove([False, False, False])  # no waves gives no sound
LEGAL_WAVES.remove([False, False, True])  # only saw gives no sound (bug?)

notes = {'sine':[],'saw':[],'square':[]}

# The samples are 8bit signed, from -127 to +127
# so the max amplitude of a sample is 127
max_sample = 2**(BITRATE - 1) - 1

def wave_sine(freq, time):
    """Generates a single sine wave sample"""

    s = math.sin(2*math.pi*freq*time)
    return int(round(max_sample * s ))


def wave_square(freq, time):
    """Generates a single square wave sample"""

    return -max_sample if freq*time < 0.5 else max_sample


def wave_saw(freq, time):
    """Generates a single sav wave sample"""

    s = ((freq*time)*2) - 1
    return int(round(max_sample * s)) 

# IMPORTANT
def generate_sample(frequency, volume=1.0, wavetype=None):
    """Generates a sample of a specific frequency and wavetype"""
    if wavetype is None:
        wavetype = wave_square

    sample_count = int(round(SAMPLERATE/frequency))

    buf = numpy.zeros((sample_count, 2), dtype = numpy.int8)

    for s in range(sample_count):
        t = float(s)/SAMPLERATE # Time index
        buf[s][0] = wavetype(frequency, t)
        buf[s][1] = buf[s][0] # Copy to stero channel

    sound = pygame.sndarray.make_sound(buf)
    sound.set_volume(volume) # Set the volume to balance sounds

    return sound

# generate samples
for f in [
        261.626,
        277.183,
        293.665,
        311.127,
        329.628,
        349.228,
        369.994,
        391.995,
        415.305,
        440.000,
        466.164,
        493.883,
        523.251
    ]:
    notes['sine'] += [generate_sample(f, volume=volume['sine'], wavetype=wave_sine)]
    notes['saw'] += [generate_sample(f, volume=volume['saw'], wavetype=wave_saw)]
    notes['square'] += [generate_sample(f, volume=volume['square'], wavetype=wave_square)]

############## /synthi constants


def parse_arguments(sysargs):
    """ Setup the command line options. """

    parser = argparse.ArgumentParser(description=DESCRIPTION)
    parser.add_argument('-p', '--piano', default='piano')
    parser.add_argument('-d', '--drums', default='drums2')

    return parser.parse_args(sysargs)


def natural_sort_key(s, _nsre=re.compile('([0-9]+)')):
    return [int(text) if text.isdigit() else text.lower() for text in re.split(_nsre, s)]


class Container:
    """ Container is a factory for creating instruments, necessary for 
    switching to 8-bit piano """

    piano = None
    drums = None

    def __init__(self, piano_index, drums_index):
        self.drums = Drums(drums_index)
        self.create_piano(piano_index)

        signal.pause()

    def create_piano(self, piano_index):
        if sound_sets[piano_index] == '8bit':
            # synthi needs no sound_index, 0 is a dummy value
            self.piano = Synthesizer(self, 0)  
        else:
            self.piano = Piano(self, piano_index)


class Instrument:
    sounds = []
    sound_index = 0

    def __init__(self, sound_index):
        self.sound_index = sound_index
        self.load_sounds()

    def load_sounds(self):
        sounds_path = glob.glob(os.path.join(SOUND_BASEDIR, 
                                                   sound_sets[self.sound_index], "*.wav"))
        sounds_path.sort(key=natural_sort_key)
        self.sounds = [pygame.mixer.Sound(f) for f in sounds_path]  


class Drums(Instrument):

    def __init__(self, sound_index):
        super(Drums, self).__init__(sound_index)

        drumhat.on_hit(drumhat.PADS, self.handle_hit)
        drumhat.on_release(drumhat.PADS, self.handle_release)

    def handle_hit(self, event):
        # event.channel is a zero based channel index for each pad
        self.sounds[event.channel].play(loops=0)

    def handle_release(self):
        pass  


# maybe add a wrapper four outputting played sound  filename?
class Piano(Instrument):
    octave = 0
    octaves = 0 
    container = None
    mixer_settings = MIXER_NORMAL

    def __init__(self, container, sound_index):
        super(Piano, self).__init__(sound_index)

        self.container = container

        pianohat.on_note(self.handle_note)
        pianohat.on_octave_up(self.handle_octave_up)
        pianohat.on_octave_down(self.handle_octave_down)
        pianohat.on_instrument(self.handle_instrument)

        pianohat.auto_leds(True)

    def load_sounds(self):
        set_mixer(self.mixer_settings)  # reset mixer to normal mode

        super(Piano, self).load_sounds()        
        self.octaves = len(self.sounds) / 12
        self.octave = int(self.octaves / 2)   

    # could be merged with handle_hit in Drum, but that'd be obfuscating
    def handle_note(self, channel, pressed):
        channel = channel + (12 * self.octave)

        if channel < len(self.sounds) and pressed:
            self.sounds[channel].play(loops=0)

    def handle_instrument(self, channel, pressed):
        if pressed:
            self.sound_index = (self.sound_index + 1) % len(sound_sets)
            self.container.create_piano(self.sound_index)

    def handle_octave_up(self, channel, pressed):
        if pressed and self.octave < int(self.octaves) - 1:
            self.octave += 1

    def handle_octave_down(self, channel, pressed):
        if pressed and self.octave > 0:
            self.octave -= 1


class Synthesizer(Piano):
    mixer_settings = MIXER_8BIT
    wavetype_index = 0

    def __init__(self, container, sound_index):
        super(Synthesizer, self).__init__(container,  sound_index)   

    def handle_note(self, channel, pressed):
        """Handles the piano keys
        Any enabled samples are played, and *all* samples are turned off is a key is released
        """
        
        if pressed:
            # 'tis so ugly
            for i in range(3):
                if LEGAL_WAVES[self.wavetype_index][i]:
                    notes[wavetypes[i]][channel].play(loops=-1, fade_ms=ATTACK_MS)
        else:
            for t in wavetypes:
                notes[t][channel].fadeout(RELEASE_MS)

    def handle_octave_up(self, channel, pressed):
        if pressed and self.wavetype_index < len(LEGAL_WAVES) - 1:
            self.wavetype_index += 1

    def handle_octave_down(self, channel, pressed):
        if pressed and self.wavetype_index > 0:
            self.wavetype_index -= 1


def turn_off(pin):
    """ Shutdown the Raspberry Pi; the argument pin is not required
    but passed by event_detect. """

    GPIO.cleanup()
    subprocess.call(['sudo poweroff'], shell=True)


if __name__ == "__main__":
    # optional shutdown button
    GPIO.add_event_detect(24, edge=GPIO.FALLING, callback=turn_off) 
    args = parse_arguments(sys.argv[1:]) 

    container = Container(sound_sets.index(args.piano), sound_sets.index(args.drums))
